Source/JavaScriptCore/runtime/OrderedHashTable.h

```diff
@@ -228,7 +228,12 @@ class OrderedHashMap : public OrderedHashTable<MapTraits> {
             value = getValueFunctor();
             RETURN_IF_EXCEPTION(scope, { });
 
-            Helper::addImpl(globalObject, this, storage, key, value, result);
+            // FIXME: rdar://145147128
+            // Could we detect if the map is changed and only rehash if needed?
+
+            // Call to getValueFunctor can modify our state, so we need to re-check the index
+            result = Helper::find(globalObject, storageRef(), key);
+            Helper::addImpl(globalObject, this, storageRef(), key, value, result);
```

Comment:
For JSMap/Set backed by `OrderedHashTable`, you can do 
```
            if (Helper::isObsolete(storage)) {
                Helper::add(globalObject, this, storage, key, value);
                RETURN_IF_EXCEPTION(scope, { });
            } else {
                Helper::addImpl(globalObject, this, storageRef(), key, value, result);
                RETURN_IF_EXCEPTION(scope, { });
            }
```
where `Helper::isObsolete` is the way to check whether it's rehashed.